quickwit-common/src/uri.rs

```diff
@@ -83,48 +90,103 @@ impl Uri {
 
                 path.replace_range(0..1, &home_dir_path);
             }
-
-            if !path.starts_with('/') {
+            if Path::new(&path).is_relative() {
                 let current_dir = env::current_dir().context(
                     "Failed to resolve current working directory: dir does not exist or \
                      insufficient permissions.",
                 )?;
                 path = current_dir.join(path).to_string_lossy().to_string();
             }
-
             path = normalize_path(Path::new(&path))
                 .to_string_lossy()
                 .to_string();
         }
-
         Ok(Self {
             uri: format!("{}{}{}", protocol, PROTOCOL_SEPARATOR, path),
             protocol_idx: protocol.len(),
         })
     }
 
-    /// Returns the URI extension.
+    /// Constructs a [`Uri`] from a properly formatted string `<protocol>://<path>` where `path` is
+    /// normalized. Use this method exclusively for trusted input.
+    pub fn new(uri: String) -> Self {
```

Comment:
It is strange to make this function public.

Reply:
Yes. I don't like that it opens the door for shooting yourself in the foot. At the same time, sometimes you know the input is trustworthy (deserialization from metastore), and that's really what you want to use. It's also very convenient for writing less verbose unit tests.

Reply:
ok, so at least there is one useful use case. 
For unit tests, we can make it available only for that.

Reply:
I don't know if we can do it across crates of the same workspace. Last time I tried I failed.